packages/tpetra/core/src/Tpetra_Distributor.hpp

```diff
@@ -261,7 +268,21 @@ namespace Tpetra {
     ///
     /// \pre No outstanding communication requests.
     ///   (We could check, but see GitHub Issue #1303.)
-    virtual ~Distributor () = default;
+    ~Distributor () {
+      for (int i = 0; i<persistentRequests_.size(); i++) {
+        Teuchos::RCP<Teuchos::CommRequest<int> > request = persistentRequests_[i];
+        Teuchos::RCP<Teuchos::MpiCommRequestBase<int> > mpiRequest =
+          Teuchos::rcp_dynamic_cast<Teuchos::MpiCommRequestBase<int> > (request);
+        MPI_Request mR = mpiRequest->releaseRawMpiRequest();
+        const int err = MPI_Request_free(&mR);
+        if (err != MPI_SUCCESS) {
+          std::ostringstream os;
+          os << "Teuchos::MpiComm::waitAll: MPI_Request_free() failed with error \""
+             << Teuchos::mpiErrorCodeToString (err) << "\".";
+          TEUCHOS_TEST_FOR_EXCEPTION(true, std::runtime_error, os.str());
```

Comment:
Destructors shouldn't throw.  Best practice would be to ignore any MPI errors at this point.  How is user code supposed to respond to them?